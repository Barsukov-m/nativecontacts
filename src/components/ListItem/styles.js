import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  listItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 20,
    paddingVertical: 10,
  },
  contactInfo: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  contactThumbnail: {
    width: 40,
    height: 40,
    borderRadius: 40 / 2,
  },
  contactName: {
    paddingLeft: 10,
    color: '#000',
  },
});
