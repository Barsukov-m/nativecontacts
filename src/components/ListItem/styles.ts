import { StyleSheet } from 'react-native';
import themeColors from '../../constants/themeColors';

export default StyleSheet.create({
	listItem: {
		flexDirection: 'row',
		justifyContent: 'space-between',
		alignItems: 'center',
		paddingHorizontal: 20,
		paddingVertical: 10,
	},
	contactInfo: {
		flexDirection: 'row',
		alignItems: 'center',
	},
	contactThumbnail: {
		width: 40,
		height: 40,
		borderRadius: 40 / 2,
	},
	missingImage: {
		backgroundColor: themeColors.primaryDark,
		color: themeColors.textOnPrimary,
		textAlign: 'center',
		verticalAlign: 'middle',
	},
	contactName: {
		paddingLeft: 10,
		color: '#000',
	},
});
