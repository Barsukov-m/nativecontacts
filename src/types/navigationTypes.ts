import { NativeStackNavigationProp } from '@react-navigation/native-stack';

export interface ScreenProps {
	navigation: NativeStackNavigationProp<any>;
}

export interface ContactInterface {
	id?: number;
	firstName: string;
	lastName: string;
	maidenName?: string;
	age?: number;
	gender?: string;
	email?: string;
	phone: string;
	username?: string;
	birthDate?: string;
	image?: string;
	address?: {
		address?: string;
		city: string;
		state?: string;
	};
	university?: string;
	company?: {
		name: string;
	};
}
