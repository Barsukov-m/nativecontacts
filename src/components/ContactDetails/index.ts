import Header from './Header';
import Section from './Section';
import ContactInfo from './ContactInfo';
import styles from './styles';

export { Header, Section, ContactInfo, styles };
