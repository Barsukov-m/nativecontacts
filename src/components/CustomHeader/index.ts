import CustomHeader from './CustomHeader';
import LeftActions from './LeftActions';
import RightActions from './RightActions';
import styles from './styles';

export { CustomHeader, LeftActions, RightActions, styles };
