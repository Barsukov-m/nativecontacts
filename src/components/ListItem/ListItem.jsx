import { Image, Text, TouchableNativeFeedback, View } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { capitalizeFirstLetter } from '../../utils/contacts';
import styles from './styles';

const ListItem = ({ contact, navigation }) => {
  const contactName = `${capitalizeFirstLetter(
    contact.name.first
  )} ${capitalizeFirstLetter(contact.name.last)}`;

  const handleNavigation = () => {
    navigation.navigate('Details', { contact: contact });
  };

  return (
    <TouchableNativeFeedback onPress={handleNavigation}>
      <View style={styles.listItem}>
        <View style={styles.contactInfo}>
          <Image
            source={{ uri: contact.picture.thumbnail }}
            style={styles.contactThumbnail}
          />
          <Text style={styles.contactName}>{contactName}</Text>
        </View>
        <Icon name="arrow-forward" size={25} />
      </View>
    </TouchableNativeFeedback>
  );
};

export default ListItem;
