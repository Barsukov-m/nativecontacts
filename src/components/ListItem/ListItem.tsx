import { Image, Text, TouchableNativeFeedback, View } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { ScreenProps } from 'types/navigationTypes';
import { ContactInterface } from '../../types/navigationTypes';
import { getHeaderTitleByContact } from '../../utils/contactsUtils';
import styles from './styles';

interface ListItemProps extends ScreenProps {
	contact: ContactInterface;
}

const ListItem: React.FC<ListItemProps> = ({ contact, navigation }) => {
	const title = getHeaderTitleByContact(contact);

	const handleNavigation = () => {
		navigation.navigate('Details', { contact: contact });
	};

	const renderContactImage = () => {
		if (contact.image) {
			return (
				<Image
					source={{ uri: contact.image }}
					style={styles.contactThumbnail}
					testID="contact-list-image"
				/>
			);
		} else {
			return (
				<Icon
					style={{ ...styles.contactThumbnail, ...styles.missingImage }}
					name="person"
					size={20}
				/>
			);
		}
	};

	return (
		<TouchableNativeFeedback onPress={handleNavigation}>
			<View style={styles.listItem}>
				<View style={styles.contactInfo}>
					{renderContactImage()}
					<Text style={styles.contactName}>{title}</Text>
				</View>
				<Icon name="arrow-forward" size={25} />
			</View>
		</TouchableNativeFeedback>
	);
};

export default ListItem;
