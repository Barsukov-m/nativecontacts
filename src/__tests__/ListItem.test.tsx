import { render, fireEvent } from '@testing-library/react-native';
import { ListItem } from '../components/ListItem';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { ContactInterface } from 'types/navigationTypes';

// Mock the navigation
const mockNavigate = jest.fn();
const navigation: any = {
	navigate: mockNavigate,
};

jest.mock('react-native-vector-icons/MaterialIcons');

describe('<ListItem />', () => {
	const contact: ContactInterface = {
		firstName: 'Terry',
		lastName: 'Medhurst',
		phone: '+63 791 675 8914',
		image: 'https://robohash.org/hicveldicta.png',
	};

	const renderComponent = (contact: ContactInterface) =>
		render(<ListItem contact={contact} navigation={navigation} />);

	it('renders correctly and renders the contact name', () => {
		const tree = renderComponent(contact);
		const { getByText } = tree;

		expect(getByText('Terry Medhurst')).toBeDefined();

		expect(tree).toMatchSnapshot();
	});

	it('renders the contact profile image when available', () => {
		const { getByTestId } = renderComponent(contact);
		const image = getByTestId('contact-list-image');

		expect(image.props.source.uri).toBe(contact.image);
	});

	it('renders the person icon when contact profile image is not available', () => {
		const { queryByTestId } = renderComponent({
			firstName: contact.firstName,
			lastName: contact.lastName,
			phone: contact.phone,
		});

		expect(queryByTestId('contact-list-image')).not.toBeTruthy();

		expect(Icon).toHaveBeenCalledWith(
			expect.objectContaining({
				name: 'person',
			}),
			{}
		);
	});

	it('navigates to Details screen when pressed', () => {
		const { getByTestId } = renderComponent(contact);
		const listItem = getByTestId('contact-list-image');
		fireEvent.press(listItem);

		expect(mockNavigate).toHaveBeenCalledWith('Details', { contact });
	});
});
