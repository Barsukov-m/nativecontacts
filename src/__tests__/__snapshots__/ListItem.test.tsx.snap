// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> renders correctly and renders the contact name 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "paddingHorizontal": 20,
      "paddingVertical": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Image
      source={
        {
          "uri": "https://robohash.org/hicveldicta.png",
        }
      }
      style={
        {
          "borderRadius": 20,
          "height": 40,
          "width": 40,
        }
      }
      testID="contact-list-image"
    />
    <Text
      style={
        {
          "color": "#000",
          "paddingLeft": 10,
        }
      }
    >
      Terry Medhurst
    </Text>
  </View>
  Icon
</View>
`;
